docs/references/config.md
---
categories:
- reference
description: Reference documentation for Homechart's configuration
title: Config
---

{{% snippet config_format Homechart homechart %}}

## Configuration Values

### `app`

Main Homechart configurations.

{{% snippet config_key "app_adminEmailAddresses" "(recommended)" %}}

List of strings, email addresses which will have admin access to Homechart for their account.  If combined with `adminTrustedCIDRs`, it will limit the admin access granted to traffic originating from the networks specified.

**Default:** `[]`

{{% snippet config_key "app_adminTrustedCIDRs" "(recommended)" %}}

List of IPv4 and IPv6 networks.  If `adminEmailAddresses` is not specified, any account from these networks will be granted admin access to Homechart.  If `adminEmailAddresses` is specified, it will limit the admin access granted for `adminEmailAddresses` to traffic originating from the networks specified.

Should be combined with `xForwardedTrustedCIDRs` to ensure the appropriate remote IP is being checked, otherwise this may grant admin access to anyone on the internet.

**Default:**
```json
[
  "10.0.0.0/8",
  "127.0.0.1/32",
  "172.16.0.0/12",
  "192.168.0.0/16",
  "::1/128",
  "fd00::/8",
]
```

{{% snippet config_key "app_baseURL" "(recommended)" %}}

String, base URL for Homechart, mostly used by notifications.  Should match the URL users will access Homechart from.

**Default:** `"https://example.com"`

{{% snippet config_key "app_cacheTTLMinutes" %}}

Number, number of minutes to keep entries in cache.

**Default:** `15`

{{% snippet config_key "app_disableNotifications" %}}

Boolean, disables notifications from being sent.

**Default:** `false`

{{% snippet config_key "app_disableSignup" "(recommended)" %}}

Boolean, disables new account signups.  Accounts can still be created/invited under the Admin > Accounts.  Self-hosted instances should enable this after they have setup their initial account.

**Default:** `false`

{{% snippet config_key "app_disableTasks" %}}

Boolean, disable any background tasks (sending notifications, cleaning up old data) from running on this instance.

**Default:** `false`

{{% snippet config_key "app_disableWebConfig" %}}

Boolean, disables configuration via Web/HTTP.

**Default:** `false`

{{% snippet config_key "app_keepCalendarEventDays" %}}

Number, number of days to retain Calendar Events after their end date.  Setting this to 0 disables Calendar Event cleanup.

**Default:** `180`

{{% snippet config_key "app_keepChangeDays" %}}

Number, number of days to retain Change logs.  These are used for client synchronization and to populate the history on the Home screen.

**Default:** `14`

{{% snippet config_key "app_keepCookMealPlanDays" %}}

Number, number of days to retain Cook Meal Plans after their scheduled date.  Setting this to 0 disables Cook Meal Plan cleanup.

**Default:** `180`

{{% snippet config_key "app_keepDeletedDays" %}}

Number, number of days to keep deleted data (Notes, Cook Recipes, Secret Values).  Setting this to 0 disables deleted data cleanup.

**Default:** `30`

{{% snippet config_key "app_keepHealthLogDays" %}}

Number, number of days to retain Health Logs.  Setting this to 0 disables Health Log cleanup.

**Default:** `180`

{{% snippet config_key "app_keepNotesPageVersions" %}}

Number, number of Notes Page Versions to keep.

**Default:** `10`

{{% snippet config_key "app_keepPlanTasksDays" %}}

Number, number of days to keep completed tasks.

**Default:** `90`

{{% snippet config_key "app_imageMaxSizePixels" %}}

Number, max pixel size of an image (height or width).  Increase this value to show higher resolution images in Homechart, at the cost of more storage space and bandwidth usage.

**Default:** `300`

{{% snippet config_key "app_motd" %}}

String, informational message to display on the UI for all users.

**Default:** `""`

{{% snippet config_key "app_privateKeys" %}}

List of private keys to use for signing and other cryptographic operations.  Users shouldn't set this value--Homechart will generate keys at startup and store them in a database.  Users can rotate these keys by removing them from Admin > Config--Homechart will generate new keys after restarting.

**Default:** `[]`

{{% snippet config_key "app_port" %}}

Number, listening port for Homechart.  Setup port forwarding to this port to expose Homechart externally.

**Default:** `3000`

{{% snippet config_key "app_proxyAddress" %}}

String, Upstream IPv4 or IPv6 address of a trusted proxy.  See the [Single Sign-On (SSO) guide]({{< ref "/docs/guides/sso" >}}) for usage details.

**Default:** `""`

{{% snippet config_key "app_proxyHeaderEmail" %}}

String, proxy header that should be associated with an account email address.  See [Single Sign-On (SSO) guide]({{< ref "/docs/guides/sso" >}}) for usage details.

**Default:** `""`

{{% snippet config_key "app_proxyHeaderName" %}}

String, proxy header that should be associated with an account name.  See [Single Sign-On (SSO) guide]({{< ref "/docs/guides/sso" >}}) for usage details.

**Default:** `""`


{{% snippet config_key "app_rollupBudgetMonthCategoriesMonths" %}}

Number, number of months before a Budget Category month detail is rolled up into the next Budget Category month.  Setting to 0 disables category roll ups.  Not rolling up categories may impact performance.

**Default:** `96`

{{% snippet config_key "app_rollupBudgetTransactionsBalanceMonths" %}}

Number, number of months before a Budget Transaction is rolled up into a starting balance.  Setting to 0 disables balance roll ups.  Not rolling up transactions may impact performance.

**Default:** `96`

{{% snippet config_key "app_rollupBudgetTransactionsSummaryMonths" %}}

Number, number of months before a Budget Transaction is rolled up into monthly summaries.  Setting to 0 disables summary roll ups.  Not rolling up transactions may impact performance.

**Default:** `48`

{{% snippet config_key "app_sessionExpirationDefault" %}}

String, duration until non-Remember Me sessions expire. {{% config_duration %}}

**Default:** `"1h"`

{{% snippet config_key "app_sessionExpirationRemember" %}}

String, duration until Remember Me sessions expire. {{% config_duration %}}

**Default:** `"90d"`

{{% snippet config_key "app_sseTimeout" %}}

String, duration until SSE requests will be removed due to inactivity.

**Default:** `"10m"`

{{% snippet config_key "app_tlsCertificate" "(recommended)" %}}

String, path to a SSL/TLS certificate file.  Should work for the domain in your [baseURL](#app_baseURL).  If set, along with [tlsKey](#app_tlsKey), Homechart will listen for HTTPS connections only.  HTTPS is necessary for certain Homechart functionality, you should enable HTTPS within Homechart or your reverse proxy.

**Default:** `""`

{{% snippet config_key "app_tlsKey" "(recommended)" %}}

String, path to a SSL/TLS private key file.  Should work for the domain in your [baseURL](#app_baseURL).  If set, along with [tlsCertificate](#app_tlsCertificate), Homechart will listen for HTTPS connections only.  HTTPS is necessary for certain Homechart functionality, you should enable HTTPS within Homechart or your reverse proxy.

**Default:** `""`

{{% snippet config_cli homechart yellow %}}

{{% snippet config_cache %}}

{{% snippet config_database %}}

{{% snippet config_httpClient Homechart %}}

{{% snippet config_httpServer ":3000" homechart %}}

{{% snippet config_jsonnet true %}}

### `oidc` {#oidc}

OIDC is a map of OIDC issuer names to OIDC configurations.  See [Setup Homechart SSO]({{< ref "/docs/guides/sso" >}}) for more information.

**Default:** `{}`

#### `oidc_[issuer_name]_authURL` {#oidc_authURL}

String, the endpoint for OAuth.  Can be discovered from [`oidcIssuerURL`](#oidc_oidcIssuerURL).

**Default:** `""`

#### `oidc_[issuer_name]_caCertificateBase64` {#oidc_caCertificateBase64}

String, the CA certificate used to communicate with the issuer.  Useful with private OIDC servers without public certificates.

**Default:** `""`

#### `oidc_[issuer_name]_clientID` {#oidc_clientID}

String, the ID shared with clients for OAuth requests.  Required.

**Default:** `""`

#### `oidc_[issuer_name]_clientSecret` {#oidc_clientSecret}

String, the secret shard with clients for OAuth requests.  Required.

**Default:** `""`

#### `oidc_[issuer_name]_codeChallengeMethodsSupported` {#oidc_codeChallengeMethodsSupported}

List of strings, the supported algorithms for OIDC code challenges.  Can be discovered from [`oidcIssuerURL`](#oidc_oidcIssuerURL).

**Default:** `""`

#### `oidc_[issuer_name]_displayName` {#oidc_displayName}

String, the name that will be displayed in the UI for the OIDC issuer.

**Default:** `""`

#### `oidc_[issuer_name]_icon` {#oidc_icon}

String, a URL for an icon to display for the OIDC issuer.  If this issuer has a name of `google` or `microsoft`, an icon will be provided by Homechart.

**Default:** `""`

#### `oidc_[issuer_name]_jwksKeys` {#oidc_jwksKeys}

List of strings, a list of public keys for an OIDC issuer.  Can be discovered from [`oidcIssuerURL`](#oidc_oidcIssuerURL).

**Default:** `""`

#### `oidc_[issuer_name]_jwksURI` {#oidc_jwksURI}

String, a URL for retrieving JWKS keys for an OIDC issuer.  Can be discovered from [`oidcIssuerURL`](#oidc_oidcIssuerURL).

**Default:** `""`

#### `oidc_[issuer_name]_issuerURL` {#oidc_issuerURL}

String, the issuer URL to be compared with the `issuer` of the token.  Can be discovered from [`oidcIssuerURL`](#oidc_oidcIssuerURL).

**Default:** `""`

#### `oidc_[issuer_name]_oidcIssuerURL` {#oidc_oidcIssuerURL}

String, URL to retrieve issuer configurations.  Should be the base path without `/.well-known/openid-configuration`, i.e. `https://accounts.google.com`.

**Default:** `""`

#### `oidc_[issuer_name]_scopes` {#oidc_scopes}

List of strings, a list of scopes to request.  If empty, will be set to "openid" and "email".  Homechart requires `sub` and `email` claims in an ID token to successfully authenticate a user using OIDC.

**Default:** `""`

#### `oidc_[issuer_name]_tokenURL` {#oidc_tokenURL}

String, the URL to retrieve tokens.  Can be discovered from [`oidcIssuerURL`](#oidc_oidcIssuerURL).

**Default:** `""`

#### `oidc_[issuer_name]_userInfoEndpoint` {#oidc_userInfoEndpoint}

String, the URL to retrieve user info.  Can be discovered from [`oidcIssuerURL`](#oidc_oidcIssuerURL).

**Default:** `""`

### `smtp` {#smtp}

Homechart can use a SMTP server to send notifications to your household members.

{{% snippet config_key "smtp_fromAddress" %}}

String, email address to send from.  Required to make SMTP work.

**Default:** `""`

{{% snippet config_key "smtp_helo" %}}

String, the HELO or EHLO value to send to the remote SMTP server.  Some SMTP services (like Google Workspace/Gmail SMTP relay) require a value that isn't `localhost` to be set here.

**Default:** `"<hostname portion of app_baseURL>"`

{{% snippet config_key "smtp_hostname" %}}

String, hostname to use when connecting to SMTP server.

**Default:** `""`

{{% snippet config_key "smtp_password" %}}

String, password to use when connecting to SMTP server.

**Default:** `""`

{{% snippet config_key "smtp_port" %}}

Number, port to use when connecting to SMTP server.

**Default:** `587`

{{% snippet config_key "smtp_replyTo" %}}

String, email address to have users send to when replying.

**Default:** `""`

{{% snippet config_key "smtp_requireTLS" %}}

Boolean, require a valid TLS certificate for all outgoing SMTP requests.

**default:** `false`

{{% snippet config_key "smtp_username" %}}

String, username to use when connecting to SMTP server.

**Default:** `""`

### `webPush` {#webpush}

Homechart can use [Web Push](https://developer.mozilla.org/en-US/docs/Web/API/Push_API) to send push notifications from Homechart to your devices.  Homechart communicates directly to web push services provided by Apple, Google, Mozilla and other standards-compliant endpoints.  Additionally, all of the data in the push notification is encrypted between your server and the client--the web push services can't read it.

You need to generate the VAPID private key to use Web Push.  Homechart will do this automatically and store the key in the database, but you can also generate a key from the command line, e.g.:

```shell
$ ./homechart_linux_amd64 generate-vapid
VEIYXV6qF_enUzycyQYdplDUgi05UM4lPh_FTzYmwX8
```

Or using a container:

```shell
$ docker run -it --rm ghcr.io/candiddev/homechart generate-vapid
VEIYXV6qF_enUzycyQYdplDUgi05UM4lPh_FTzYmwX8
```

This command will output the private key you'll use in the configuration sections below.

{{% snippet config_key "webPush_vapidPrivateKey" %}}

String, the privateKey value from running `generate-vapid`.  Will be automatically generated by Homechart if one isn't provided.

**Default:** `""`


docs/docs/guides/settings.md
---
categories:
- guide
description: Usage instructions for Settings
title: Settings
---

Settings are where you manage all of the various controls for Homechart.

## Account

Settings > Account has controls for adjusting your **Preferences**, **Components**, and **Security**.

### Preferences {#account-preferences}

You can adjust your account **Name**, **Language, Date, and Time**, and **Color and Theme** under **Preferences**.  Your account **Name** will be visible to all members of your Household.

### Hide Components {#account-hide}

Using the controls on this page, you can hide various components within Homechart.  You may not be able to un-hide items that are hidden at the [household level](#household-hide).

### Security

You can setup your **Private Key**, change your **Password**, and setup **Two-Factor Authentication** under **Security**.

#### Private Key

You need to generate a **Private Key** to use Secrets and other encryption components in Homechart.  You can generate multiple **Private Key Passphrases** to store within a safe or somewhere safe.

Homechart uses **2048-bit RSA-OAEP** to generate a Public/Private key-pair.  The **Private Key** is encrypted using **100,000 iterations of PBKDF2** using the **Private Key Passphrases** as the keys.  All encryption is performed using standard [Web Crypto APIs](https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API).

#### Delegations

Accounts can delegate access by adding **Account Delegations**.  The **Target Account** will be able to sign in as the **Source Account** via **Switch Account**.  Additional access can also be shared, such as **Health** to view/add Health data.  The delegation can also be configured to only allow the **Target Account** to change or remove the delegation--perfect for child accounts.

## Notifications

Settings > Notifications has controls for enabling/disabling different types of Notifications from Homechart.

## Sessions

Settings > Sessions lets you delete Sessions, **Clear All Sessions**, and add new sessions/generate API keys for programmatic usage of Homechart.

### Permissions {#session-permissions}

Homechart permissions can be set per-session, allowing you to have very granular access on each device you sign into.  You can use session permissions to control access on shared devices, such as tablets or other kiosks, to ensure they can only view certain things in Homechart.

## Households

Settings > Households is where you Add, Delete, and Leave households within Homechart.  You can be part of many Households, like one for your immediate family and one for your extended family.  To use Household features within Homechart, you must be part of at least one Household.

One Household will always be your **Primary Household**.  This is the default Household you'll use when you create new Household items.  You can change this by tapping the checkbox in the **Primary** column.

### Household Settings

Tapping a Household under Settings > Households shows you controls for adjusting **Preferences**, **Members**, and **Components**.

#### Preferences {#household-preferences}

You can adjust your household **Name**, **Colors**, and **Currency** under **Preferences**.

### Members

You can add, remove, and change the permissions of household members under **Members**.

### OIDC

You can add, remove, and change OIDC clients for your household.  See [Homechart as an Identity Provider]({{% ref "/docs/guides/sso#provider" %}}) for more details.


go/config/config.go
// Package config contains configuration structs for Homechart.
package config

import (
	"context"

	"github.com/candiddev/shared/go/cli"
	"github.com/candiddev/shared/go/config"
	"github.com/candiddev/shared/go/cryptolib"
	"github.com/candiddev/shared/go/errs"
	"github.com/candiddev/shared/go/logger"
	"github.com/candiddev/shared/go/notify"
	"github.com/candiddev/shared/go/paddle"
	"github.com/candiddev/shared/go/postgresql"
)

// Config contains all of the application configuration settings.
type Config struct {
	App        App                  `json:"app"`
	CLI        cli.Config           `json:"cli"`
	OAuth      OAuth                `json:"oauth,omitempty"`
	OIDC       OIDC                 `json:"oidc,omitempty"`
	Paddle     paddle.Config        `json:"paddle,omitempty"`
	PostgreSQL postgresql.Config    `json:"postgresql"`
	SMTP       notify.SMTP          `json:"smtp"`
	Tracing    logger.TracingConfig `json:"tracing,omitempty"`
	WebPush    notify.WebPush       `json:"webPush"`
}

// App contains config options.
type App struct {
	DisableTasks                          bool                                        `json:"disableTasks,omitempty"`
	IgnorePaymentFailures                 bool                                        `json:"ignorePaymentFailures,omitempty"`
	Demo                                  bool                                        `json:"demo,omitempty"`
	SignupDisabled                        bool                                        `json:"signupDisabled"`
	TestNotifier                          bool                                        `json:"-"`
	CacheTTLMinutes                       int                                         `json:"cacheTTLMinutes"`
	KeepCalendarEventDays                 int                                         `json:"keepCalendarEventDays"`
	KeepCookMealPlanDays                  int                                         `json:"keepCookMealPlanDays"`
	KeepDeletedDays                       int                                         `json:"keepDeletedDays"`
	KeepExpiredAuthHouseholdDays          int                                         `json:"keepExpiredAuthHouseholdDays,omitempty"`
	KeepHealthLogDays                     int                                         `json:"keepHealthLogDays"`
	KeepInactiveAuthAccountDays           int                                         `json:"keepInactiveAuthAccountDays,omitempty"`
	KeepNotesPageVersions                 int                                         `json:"keepNotesPageVersions"`
	KeepPlanTaskDays                      int                                         `json:"keepPlanTaskDays"`
	Port                                  int                                         `json:"port"`
	RollupBudgetTransactionsBalanceMonths int                                         `json:"rollupBudgetTransactionsBalanceMonths"`
	RollupBudgetTransactionsSummaryMonths int                                         `json:"rollupBudgetTransactionsSummaryMonths"`
	SessionExpirationDefaultSeconds       int                                         `json:"sessionExpirationDefaultSeconds"`
	SessionExpirationRememberSeconds      int                                         `json:"sessionExpirationRememberSeconds"`
	TrialDays                             int                                         `json:"trialDays,omitempty"`
	BaseURL                               string                                      `json:"baseURL"`
	CacheControl                          string                                      `json:"cacheControl,omitempty"`
	ContactFeedback                       string                                      `json:"contactFeedback,omitempty"`
	ContactSupport                        string                                      `json:"contactSupport,omitempty"`
	CounterTimeZone                       string                                      `json:"counterTimeZone,omitempty"`
	CloudEndpoint                         string                                      `json:"cloudEndpoint"`
	CloudJWT                              string                                      `json:"cloudJWT,omitempty"`        //nolint: tagliatelle
	CloudPrivateKey                       cryptolib.Key[cryptolib.KeyProviderPrivate] `json:"cloudPrivateKey,omitempty"` //nolint: tagliatelle
	CloudPublicKey                        cryptolib.Key[cryptolib.KeyProviderPublic]  `json:"-"`
	GTMID                                 string                                      `json:"gtmID,omitempty"`
	MOTD                                  string                                      `json:"motd"`
	RateLimiterKey                        string                                      `json:"rateLimiterKey,omitempty"`
	RateLimiterRate                       string                                      `json:"rateLimiterRate"`
	ProxyHeaderEmail                      string                                      `json:"proxyHeaderEmail"`
	ProxyHeaderName                       string                                      `json:"proxyHeaderName"`
	ProxyAddress                          string                                      `json:"proxyAddress"`
	SystemConfigKey                       string                                      `json:"systemConfigKey,omitempty"`
	SystemHealthKey                       string                                      `json:"systemHealthKey,omitempty"`
	SystemMetricsKey                      string                                      `json:"systemMetricsKey,omitempty"`
	SystemPprofKey                        string                                      `json:"systemPprofKey,omitempty"`
	SystemStopKey                         string                                      `json:"systemStopKey,omitempty"`
	TLSCertificate                        string                                      `json:"tlsCertificate"`
	TLSKey                                string                                      `json:"tlsKey"`
	UIDir                                 string                                      `json:"uiDir,omitempty"`
	UIHost                                string                                      `json:"uiHost,omitempty"`
	AdminEmailAddresses                   []string                                    `json:"adminEmailAddresses"`
	FeatureVotes                          []string                                    `json:"featureVotes,omitempty"`
}

// OAuth contains config options.
type OAuth struct {
	TestID  string `json:"testID,omitempty"`
	TestKey string `json:"testKey,omitempty"`
}

// OIDC contains config options.
type OIDC struct {
	GoogleClientID     string `json:"googleClientID,omitempty"`
	GoogleClientSecret string `json:"googleClientSecret,omitempty"`
}

// Default generates a default configuration.
func Default() *Config {
	return &Config{
		App: App{
			BaseURL:                               "https://web.homechart.app",
			CacheControl:                          "public, max-age=31536000, immutable",
			CacheTTLMinutes:                       15,
			CloudEndpoint:                         "https://web.homechart.app",
			KeepExpiredAuthHouseholdDays:          180,
			KeepInactiveAuthAccountDays:           180,
			KeepCalendarEventDays:                 180,
			KeepCookMealPlanDays:                  180,
			KeepDeletedDays:                       30,
			KeepHealthLogDays:                     180,
			KeepNotesPageVersions:                 10,
			KeepPlanTaskDays:                      180,
			Port:                                  3000,
			RateLimiterRate:                       "10-M",
			RollupBudgetTransactionsBalanceMonths: 48,
			RollupBudgetTransactionsSummaryMonths: 12,
			SessionExpirationDefaultSeconds:       3600,
			SessionExpirationRememberSeconds:      7776000,
			TrialDays:                             14,
		},
		PostgreSQL: postgresql.Config{
			Hostname:           "localhost",
			MaxConnections:     25,
			MaxIdleConnections: 5,
			MaxLifetimeMinutes: 5,
			Port:               5432,
			SSLMode:            "disable",
		},
		SMTP: notify.SMTP{
			NoEmailDomains: []string{"example.com"},
			Port:           587,
		},
	}
}

func (c *Config) CLIConfig() *cli.Config {
	return &c.CLI
}

func (c *Config) Parse(ctx context.Context, configArgs []string) errs.Err {
	return config.Parse(ctx, c, configArgs, "HOMECHART", c.CLI.ConfigPath)
}
